dev/com.ibm.ws.app.manager.wab/src/com/ibm/ws/app/manager/wab/internal/WABInstaller.java

```diff
@@ -568,9 +577,33 @@ protected boolean uninstallFromWebContainer(WAB wab) {
                             wabGroupsLock.unlock();
                         }
                         if (wab.getCreatedApplicationInfo()) {
+                            // system WABs get uninstalled here
                             deployedApp.uninstallApp(wab);
                         } else {
-                            deployedApp.uninstallModule(wab);
+                            final CountDownLatch latch = new CountDownLatch(1);
+                            if (ctx.getBundle(0).getState() == Bundle.STOPPING) {
+                                // if shutting down, launch uninstallModule asynchronously and wait up
+                                // to 30 secs. This avoids blocking server shutdown unduly.
+                                executorService.getService().submit(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        deployedApp.uninstallModule(wab);
+                                        latch.countDown();
+                                    }
+                                }, Boolean.TRUE);
+                                try {
+                                    // If a quiesce has not been initiated then a force stop was called so
+                                    // continue without waiting. Otherwise give the uninstallModule thread
+                                    // time to unwind before proceeding with WAB removal.
+                                    if (quiesceStarted.get()) {
+                                        latch.await(30, TimeUnit.SECONDS);
+                                    }
+                                } catch (InterruptedException e) {
+                                    // Continue with wab uninstall on this thread.
```

Comment:
Typically the best practice is to do the following when you catch the InterruptedException and don't want to propagate any errors:

```
Thread.currentThread().interrupt();
```
This is to keep the thread interrupted so that anything else doing work has to exit quickly also.